File: backend/src/main/java/com/vacanza/backend/dto/response/GamificationProfileDTO.java
package com.vacanza.backend.dto.response;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.List;

/**
 * Gamification profile response matching the frontend contract.
 * All display texts are built by the backend — no hardcoding on frontend.
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class GamificationProfileDTO {

    private String roleText; // "Urban Adventurer"
    private String levelText; // "Level 5"
    private int xpProgressPercent; // 0-100
    private int xpToNextLevel; // XP needed for next level
    private int totalXp;
    private String badgesSectionTitle; // "Achievement Badges"

    private List<StatDTO> stats;
    private List<BadgeDTO> badges;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class StatDTO {
        private String label; // "Places", "Badges", "Days"
        private long value;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class BadgeDTO {
        private Long id;
        private String title; // "Explorer", "Foodie"
        private String key; // "explorer" — frontend maps to icon
        private boolean earned;
    }
}

File: mobile/lib/features/gamification/data/models/gamification_profile_dto.dart
import 'badge_dto.dart';
import 'stat_dto.dart';

/// Top-level response from `GET /users/me/gamification`.
///
/// All fields have safe defaults. [xpProgressPercent] is clamped to 0–100.
class GamificationProfileDto {
  final String roleText;
  final String levelText;
  final int xpProgressPercent;
  final int xpToNextLevel;
  final int totalXp;
  final String badgesSectionTitle;
  final List<StatDto> stats;
  final List<BadgeDto> badges;

  const GamificationProfileDto({
    required this.roleText,
    required this.levelText,
    required this.xpProgressPercent,
    required this.xpToNextLevel,
    required this.totalXp,
    required this.badgesSectionTitle,
    required this.stats,
    required this.badges,
  });

  factory GamificationProfileDto.fromJson(Map<String, dynamic> json) {
    final rawStats = json['stats'];
    final rawBadges = json['badges'];

    return GamificationProfileDto(
      roleText: (json['roleText'] ?? '').toString(),
      levelText: (json['levelText'] ?? '').toString(),
      xpProgressPercent:
          ((json['xpProgressPercent'] as num?)?.toInt() ?? 0).clamp(0, 100),
      xpToNextLevel: (json['xpToNextLevel'] as num?)?.toInt() ?? 0,
      totalXp: (json['totalXp'] as num?)?.toInt() ?? 0,
      badgesSectionTitle: (json['badgesSectionTitle'] ?? '').toString(),
      stats: rawStats is List
          ? rawStats
              .map((e) => StatDto.fromJson(e as Map<String, dynamic>))
              .toList()
          : const [],
      badges: rawBadges is List
          ? rawBadges
              .map((e) => BadgeDto.fromJson(e as Map<String, dynamic>))
              .toList()
          : const [],
    );
  }

  @override
  String toString() =>
      'GamificationProfileDto(role: $roleText, level: $levelText, '
      'xp: $totalXp, progress: $xpProgressPercent%, '
      'stats: ${stats.length}, badges: ${badges.length})';
}
